main.py

```diff
@@ -1,53 +1,223 @@
+    if val_data is not None:
+        history = pd.DataFrame(columns=["loss", "accuracy", "val_loss", "val_accuracy"])
+    else:
+        history = pd.DataFrame(columns=["loss", "accuracy"])
+    for epoch in range(n_epochs):
+        running_loss = 0.0
+        correct = 0
+        total = 0
+        for i, data in enumerate(train_data, 0):
+            inputs, labels = data
+            total += labels.size(0)
+
+            optimizer.zero_grad()
+            output = model(inputs)
+            loss = criterion(output, labels)
+            loss.backward()
+            optimizer.step()
+
+            # calculate accuracy
+            _, preds = torch.max(output, 1)
+            for pred, label in zip(preds, labels):
+                if pred == label:
+                    correct += 1
+            running_loss += loss.item()
+        running_loss = running_loss / len(train_data)
+        running_accuracy = correct / total
+
+        if val_data is not None:
+            val_loss, val_accuracy = evaluate(model, val_data)
+            history = pd.concat(
+                [
+                    history,
+                    pd.DataFrame(
+                        {"loss": running_loss, "accuracy": running_accuracy, "val_loss": val_loss, "val_accuracy": val_accuracy},
+                        index=[epoch],
+                    ),
+                ]
+            )
+            print(
+                f"[{epoch + 1}] loss: {running_loss:.3f} accuracy: {running_accuracy:.3f}"
+                + f" val_loss: {val_loss:.3f} val_accuracy: {val_accuracy:.3f}"
+            )
+        else:
+            history = pd.concat([history, pd.DataFrame({"loss": running_loss, "accuracy": running_accuracy}, index=[epoch])])
+            print(f"[{epoch + 1}] loss: {running_loss:.3f} accuracy: {running_accuracy:.3f}")
+
+    print("Finished Training")
+    return model, history
+
+
+def evaluate(model, test_dataloader: DataLoader):
+    """
+    Evaluates the performance of a given model
+
+    Args:
+        model (torch.nn.Module): The trained model to be evaluated.
+        test_dataloader (DataLoader): The test dataset loader containing notes and labels in batches.
+
+    Returns:
+        Returns:
+        tuple: A tuple containing the evaluation results.
+            - val_loss (float): The average loss of the model on the test dataset.
+            - val_accuracy (float): The accuracy of the model on the test dataset, as a percentage.
+    """
+    # get test data
+    criterion = nn.CrossEntropyLoss
+    loss_sum = 0
+    correct = 0
+    total = 0
+    with torch.no_grad():
+        for data in test_dataloader:
+            notes, labels = data
+            total += labels.size(0)
+            out = model(notes)
+            loss = criterion(out, labels)
+            loss_sum += loss.item()
+            _, preds = torch.max(out, 1)
+            for pred, label in zip(preds, labels):
+                if pred == label:
+                    correct += 1
```

Comment:
```suggestion
            correct += (pred == label).sum()
```

It's almost always better to not have fors and ifs :)